org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/StackFramePresentationProvider.java

```diff
@@ -0,0 +1,214 @@
+	public static final FilterManager CUSTOM_STACK_FRAMES = new FilterManager(IJDIPreferencesConstants.PREF_ACTIVE_CUSTOM_FRAME_FILTER_LIST, IJDIPreferencesConstants.PREF_INACTIVE_CUSTOM_FRAME_FILTER_LIST);
+
+	/**
+	 * Class to decide if a particular class name is part of a list of classes and list of packages.
+	 */
+	static class Filters {
+		private final String[] filters;
+
+		Filters(String[] filters) {
+			this.filters = filters;
+		}
+
+		boolean match(String fqcName) {
+			for (String filter : filters) {
+				if (filter.endsWith("*")) { //$NON-NLS-1$
+					if (fqcName.startsWith(filter.substring(0, filter.length() - 1))) {
+						return true;
+					}
+				} else {
+					if (filter.equals(fqcName)) {
+						return true;
+					}
+				}
+			}
+			return false;
+		}
+	}
+
+	private Filters platform;
+	private Filters custom;
+	private final IPreferenceStore store;
+	private boolean collapseStackFrames;
+
+	public StackFramePresentationProvider(IPreferenceStore store) {
+		this.store = store;
+		platform = createActivePlatformFilters(store);
+		custom = createActiveCustomFilters(store);
+		store.addPropertyChangeListener(this);
+		collapseStackFrames = store.getBoolean(IJDIPreferencesConstants.PREF_COLLAPSE_STACK_FRAMES);
+	}
+
+	/**
+	 * Create a {@link Filters} object to decide if a class is part of the 'platform'. The platform definition is stored in the
+	 * {@link IPreferenceStore}. By default, this is the classes provided by the JVM.
+	 *
+	 */
+	private Filters createActivePlatformFilters(IPreferenceStore store) {
+		return new Filters(PLATFORM_STACK_FRAMES.getActiveList(store));
+	}
+
+	/**
+	 * Create a {@link Filters} object to decide if a class is considered part of a custom, very important layer, which needs to be highlighted. This
+	 * definition is stored in the {@link IPreferenceStore}. By default, this is an empty list.
+	 */
+	private Filters createActiveCustomFilters(IPreferenceStore store) {
+		return new Filters(CUSTOM_STACK_FRAMES.getActiveList(store));
+	}
+
+	public StackFramePresentationProvider() {
+		this(JDIDebugUIPlugin.getDefault().getPreferenceStore());
+	}
+
+	/**
+	 * @return the category specific image for the stack frame, or null, if there is no one defined.
+	 */
+	public ImageDescriptor getStackFrameImage(IJavaStackFrame frame) {
+		var category = getCategory(frame);
+		if (category != null) {
+			switch (category) {
+				case LIBRARY:
+				case SYNTHETIC:
+				case PLATFORM:
+					return JavaPluginImages.DESC_OBJS_JAR;
+				default:
+					break;
+			}
+		}
+		return null;
+	}
+
+	/**
+	 * @return the {@link IJavaStackFrame.Category} which matches the rules.
+	 */
+	public IJavaStackFrame.Category getCategory(IJavaStackFrame frame) {
+		if (frame.getCategory() != null) {
+			return frame.getCategory();
+		}
+		IJavaStackFrame.Category result = Category.UNKNOWN;
+		try {
+			result = categorize(frame);
+		} catch (DebugException e) {
+			JDIDebugUIPlugin.log(e);
+		}
+		frame.setCategory(result);
```

Comment:
why does a getter getCategory **set** something? that is unexpected

Reply:
so if setCategory is only used as cache please move the caching implementation into the getter to avoid the setter. However it feels like the implementation is wrong in means of it will not update the cached value if the options change.

Reply:
The setter was there, because all the settings were managed by JDIDebugUIPlugin, not by JDIDebugPlugin, partially because managing class/package filter lists are a bit easier with org.eclipse.jface.preference.IPreferenceStore and org.eclipse.jdt.internal.ui.filtertable.FilterManager - those are UI classes, not reachable from the backend.
 But I could rewrote it, the base functionality seems to be working - I still have to refactor the settings page and the couple of tests, but that will come in the other PR.